remoteappmanager/cli/remoteapprest/__main__.py

```diff
@@ -0,0 +1,210 @@
+            f.write("{}\n".format(self.url))
+            f.write("{}\n".format(self.username))
+            for k, v in self.cookies.items():
+                f.write("{}={}\n".format(k, v))
+
+
+class RemoteAppRestContext:
+    """The click context passed around."""
+    credentials_file = None
+    credentials = None
+
+
+@click.group()
+@click.option("--credentials-file",
+              default=os.path.expanduser("~/.remoteapprest"),
+              help="Specify a different credentials file.")
+@click.pass_context
+def cli(ctx, credentials_file):
+    """Command line interface to start, stop, and inquire applications
+    on the remote application server."""
+    ctx.obj = RemoteAppRestContext()
+    ctx.obj.credentials_file = credentials_file
+    try:
+        ctx.obj.credentials = Credentials.from_file(credentials_file)
+    except IOError:
+        ctx.obj.credentials = None
+
+
+@cli.command()
+@click.argument("url")
+@click.option("--username", prompt=True)
+@click.option('--password',
+              prompt=True,
+              confirmation_prompt=False,
+              hide_input=True)
+@click.pass_context
+def login(ctx, url, username, password):
+    """Performs login on the remote server at the specified URL."""
+    login_url = urljoin(url, "/hub/login")
+
+    payload = {"username": username, "password": password}
+
+    # Unfortunately, jupyterhub handles the afterlogin with an immediate
+    # redirection, meaning that we have to check for a 302 and prevent
+    # redirection in order to capture the cookies.
+    try:
+        response = requests.post(login_url, payload, verify=False,
+                                 allow_redirects=False)
+    except Exception as e:
+        print("Could not perform request. {}".format(e), file=sys.stderr)
+        sys.exit(1)
+
+    if response.status_code == 302:
+        cookies_dict = requests.utils.dict_from_cookiejar(response.cookies)
+        cred = Credentials(url, username, cookies_dict)
+        cred.write(ctx.obj.credentials_file)
+    else:
+        print("Failed to perform login. Server replied with error: {}".format(
+            response.status_code), file=sys.stderr)
+        sys.exit(1)
+
+# -------------------------------------------------------------------------
+
+
+@cli.group()
+@click.pass_context
+def app(ctx):
+    """Various subcommands to inquire the remote server."""
+    if ctx.obj.credentials is None:
+        raise click.ClickException("Missing credentials. "
+                                   "Use the login command to authenticate.")
+
+
+@app.command()
+@click.pass_context
+def available(ctx):
+    """Shows the available applications."""
+    cred = ctx.obj.credentials
+    url, username, cookies = cred.url, cred.username, cred.cookies
+
+    request_url = urljoin(url,
+                          "/user/{}/api/v1/applications/".format(username))
+    response = requests.get(request_url, cookies=cookies, verify=False)
```

Comment:
Do we need to check the response code here?  For example the cookie may have expired etc...